pandas/core/reshape/reshape.py

```diff
@@ -470,8 +470,15 @@ def factorize(index):
         if is_extension_array_dtype(dtype):
             arr = dtype.construct_array_type()
             new_values = arr._concat_same_type([
-                col for _, col in frame.iteritems()
+                col._values for _, col in frame.iteritems()
             ])
+            # final take to get the order correct.
+            # idx is an indexer like
+            # [c0r0, c1r0, c2r0, ...,
+            #  c0r1, c1r1, c241, ...]
+            idx = np.arange(N * K).reshape(K, N).T.ravel()
```

Comment:
why is this needed here?

Reply:
Had the order backwards on master :/ Without it, we were returning

```
In [6]: df.stack()
Out[6]:
0  A    a
   B    b
1  A    a
   B    b
dtype: category
Categories (2, object): [a, b]
```

instead of the correct

```
In [6]: df.stack()
Out[6]:
0  A    a
   B    a
1  A    b
   B    b
dtype: category
Categories (2, object): [a, b]
```